lib/puma/plugin/tmp_restart.rb

```diff
@@ -10,6 +10,7 @@ def start(launcher)
 
     # If we can't write to the path, then just don't bother with this plugin
     begin
+      Dir.mkdir("tmp") unless Dir.exist?("tmp")
```

Comment:
Are we adding this purely so that the tests work?

Reply:
point being, can't we do this in the test?

Reply:
No.  Or, maybe on Windows as it's currently tested in Appveyor.  We're doing it because compiling creates the directory.  It won't work without compiling.  If it's a 'sample', it shouldn't require that the `tmp` folder exist.

Reply:
I really don't want to modify the behavior of a years-old plugin just we can get our appveyor tests to pass (which already have an expiration date due to Github CI)

Reply:
I'm think more of someone wanting use it after installing the gem.  Does the `tmp` folder exists then?

Reply:
When it comes to touching files on people's deployments, I really don't want to do it. Most people's deploys are an absolute house of cards. If we don't absolutely need to do this and no one has requested it in issues, I don't want to do it.

Reply:
To even use the plugin, it will touch files on people's deployments.

Maybe we should add a rescue with a message that the folder must exist?

Reply:
what if Puma doesn't have permissions to create the folder? now this plugin (silently) dies where before it would continue to work. People may have deploy scripts out there that work around this and a different user or privileged script may create the folder after Puma boots.

My point is that this is actually a major API change that's unneccessary for this PR is trying to accomplish, regardless of where or not it's a good idea to do this.

Reply:
Like I feel like we can almost guarantee someone will create an issue about this if we deploy this in 4.2.

Reply:
> Like I feel like we can almost guarantee someone will create an issue

Bring it on!

> a major API change

In an 'example' plugin file?

Anyway, this seems to be a 'hot button' issue, so I moved the statement in question to the test....

Reply:
@nateberkopec @dentarg

I just did a fresh install of puma in 2.6.4, and set up an existing server script (for viewing docs locally) to use the plugin.

With no 'tmp' folder, there was no message, but it didn't create `restart.txt` either, so the plugin wasn't loaded.

When I added the `tmp` folder, it works.

Hence, the current plugin will not load without an existing 'tmp' folder.

Maybe one of your could try it?

Reply:
I think @nateberkopec is right here @MSP-Greg 

The [docs are pretty clear](https://github.com/puma/puma/blob/v4.1.1/docs/plugins.md)

> tmp_restart: Restarts the server if the file `tmp/restart.txt` is touched

If anyone wants to use this plugin, they will create the directory. So I think the tests should do that. 

(I can also add that `tmp/` exists by default on Heroku, as their build system puts stuff there.)

> In an 'example' plugin file?

`lib/puma/plugin/tmp_restart.rb` doesn't look like an example to me, it is the plugin, isn't it?

Reply:
@dentarg

Thanks.

> `lib/puma/plugin/tmp_restart.rb` doesn't look like an example to me, it is the plugin, isn't it?

Maybe we'll say a simple example of the hooks/methods needed to create a simple plugin?

To use it, one needs to create a config file with the line `plugin 'tmp_restart'` and add it to the puma command arguments, along with `-C`.

The current plugin is using the equivalent of `"#{Dir.pwd}/tmp"`, which may or may not exist...